Implement companion overloading

LPTK opened the pull request:
This PR adds some preliminary amount of overloading, for example between a class and its companion module.

The PR ended up being quite the yak-shave, as the required changes affected many parts of the compiler, with cascading required changes. It required contributions from four different people!

@FlandiaYingman attempted an ill-fated symbol refactoring, but bailed out mid-way and will make another separate attempt later.

---

- [x] Remove the `describe` and `expand` hacks
- [x] @LPTK Improve lowering error message when symbol is not found
- [ ] [now [moot](https://github.com/hkust-taco/mlscript/pull/331/commits/9c608c771e1501655a3a5964f4261e6c8cfa83c8)] @CAG2Mark Improve the `FIXME: DisambBlockMemberSymbol workaround` comments, esp. in Lifer
- [x] @FlandiaYingman Remove `Unexpanded ` noise

LPTK committed:
WW

LPTK committed:
WW

LPTK committed:
WW

LPTK committed:
WW

LPTK committed:
WW

LPTK committed:
WW

LPTK committed:
WW

LPTK committed:
WW

LPTK committed:
WW

LPTK committed:
WW

FlandiaYingman committed:
MM

LPTK committed:
Exhibit symbol problem

LPTK committed:
WIP fix module symbol problem (there are still other problems)

FlandiaYingman committed:
Fix funny `Foo.class` problem

FlandiaYingman committed:
Fix moduleful checks for overloaded terms

FlandiaYingman committed:
Fix funny rewriting

LPTK committed:
WW

LPTK committed:
WW

LPTK committed:
WW

LPTK committed:
WW

LPTK committed:
WW

FlandiaYingman committed:
MM

LPTK committed:
WW (?!)

# Conflicts:
#	hkmc2/shared/src/test/mlscript/HkScratch5.mls

FlandiaYingman committed:
MM2

LPTK committed:
WW run tests

LPTK committed:
WW

LPTK committed:
WW

LPTK committed:
WW

LPTK committed:
WW

LPTK committed:
WW

LPTK committed:
WW

LPTK committed:
WW

LPTK committed:
WWW

LPTK committed:
WW

LPTK committed:
WW

LPTK committed:
WW

LPTK committed:
WIP refactor + cleanup (making lifter tests fail)

LPTK committed:
Merge remote-tracking branch 'hkust-taco/hkmc2' into hkmc2-companion-overloading

# Conflicts:
#	hkmc2/shared/src/main/scala/hkmc2/utils/Scope.scala
#	hkmc2/shared/src/test/mlscript-compile/Runtime.mjs
#	hkmc2/shared/src/test/mlscript/codegen/ClassInClass.mls
#	hkmc2/shared/src/test/mlscript/codegen/SanityChecks.mls
#	hkmc2/shared/src/test/mlscript/lifter/ClassInFun.mls
#	hkmc2/shared/src/test/mlscript/lifter/DefnsInClass.mls
#	hkmc2/shared/src/test/mlscript/lifter/ModulesObjects.mls

LPTK committed:
Merge remote-tracking branch 'origin/hkmc2' into hkmc2-companion-overloading

# Conflicts:
#	hkmc2/shared/src/main/scala/hkmc2/semantics/Resolver.scala

LPTK committed:
Partially revert "WIP refactor + cleanup (making lifter tests fail)"

This reverts commit 3f60c7bdc462390f6f896193c2d03d603a1827a7.

# Conflicts:
#	hkmc2/shared/src/main/scala/hkmc2/codegen/js/JSBuilder.scala
#	hkmc2/shared/src/test/mlscript/HkScratch.mls
#	hkmc2/shared/src/test/mlscript/lifter/DefnsInClass.mls
#	hkmc2/shared/src/test/mlscript/lifter/ModulesObjects.mls

LPTK committed:
Add test file

LPTK committed:
More tests

LPTK committed:
Minor

LPTK committed:
Merge branch 'hkmc2' into hkmc2-companion-overloading

# Conflicts:
#	hkmc2/shared/src/main/scala/hkmc2/semantics/Elaborator.scala

CAG2Mark committed:
fixed some things

FlandiaYingman committed:
WIP: Resolver and Overloaded Term|Type

CAG2Mark committed:
fix

CAG2Mark committed:
update test, remove unused cases, document some things

LPTK commented on hkmc2/shared/src/main/scala/hkmc2/semantics/Elaborator.scala:
```diff
@@ -544,7 +544,9 @@ extends Importer:
-        if !inAppPrefix && ms.isParameterizedMethod && !preTrm.symbol.exists(_.isModuleful) =>
+        // FIXME: move the check to resolver becuase preTrm's symbol may
+        // not be resolved yet.
+        if !inAppPrefix && ms.isParameterizedMethod && preTrm.symbol.exists(_.existsNonModuleful) =>
```
👍

LPTK commented on hkmc2/shared/src/main/scala/hkmc2/semantics/Symbol.scala:
```diff
@@ -28,9 +28,24 @@ abstract class Symbol(using State) extends Located:
+  def existsNonModuleful: Bool = this match
+    case mod: ModuleSymbol => !(mod.tree.k is Mod)
+    case mem: BlockMemberSymbol => 
+      mem.trees.isEmpty ||
```
Maybe add a comment on why this is needed? @FlandiaYingman

FlandiaYingman replied:
OK!

FlandiaYingman replied:
BTW this comes from a limitation of `BlockMemberSymbol` itself that we might want to fix it in the future: currently all overloaded definitions share the same `BlockMemberSymbol`, but the `defn` field that lies in the `BlockMemberSymbol` is just the bottommost definition only, so one cannot reliably read the `defn` field.

Currently we fish the symbol or definition from the `trees` field of `BlockMemberSymbol` in order to find an overloaded symbol (or definition), but this approach does not work if the definition does not have a tree at all!

```fs
data class Foo(x) with
  // x internally has a val definition, but it does not have a corresponding tree!
  fun f = x
```

I think we could make `BlockMemberSymbol` not extending `MemberSymbol[Definition]`, and add a `var defns: Ls[Definition]` (or maybe a Map?) for filtering definitions...

LPTK replied:
`BlockMemberSymbol` is very loosely defined (based on a plain list of trees) because it's constructed very early, before elaboration.

The idea was to have definitions have a much more specific structure, where illegal states (ie illegal overloadings) are not representable. So storing a list of `defns` would defeat that – it would be pretty poor design. Instead, we should store a definition that precisely corresponds to the overloaded structure. This is currently somewhat (but not fully) represented in `Definition`, but the elaborator probably doesn't store the right info.

LPTK commented on hkmc2/shared/src/test/mlscript/HkScratch100.mls:
```diff
@@ -0,0 +1,70 @@
+module TypMod with
+  val whoami = "module"
+
+:todo
```
Seems like this should be `:fixme`. @FlandiaYingman

FlandiaYingman replied:
OK, fixed! But what's the semantic difference between them

LPTK replied:
I'd recommend against using `:todo`, actually. It is not specific enough. If you end up implementing the feature that needs to be done, `:todo` will remain without complaining, instead of failing.

LPTK commented on hkmc2/shared/src/test/mlscript/HkScratch100.mls:
```diff
@@ -0,0 +1,70 @@
+
+fun f: module TypMod
+
+:todo
```
A `:todo` should specify what's going wrong in a comment. What error are you expecting? @FlandiaYingman

LPTK replied:
Also, it should be something like
```suggestion
:breakme
:e // blah
```

FlandiaYingman replied:
OK, added!

LPTK commented on hkmc2/shared/src/test/mlscript/HkScratch100.mls:
```diff
@@ -0,0 +1,70 @@
+
+fun assertNonModule(m) = m
+
+// FIXME: should be resolver error
```
```suggestion
:breakme // FIXME: should be resolver error
:e
```

FlandiaYingman replied:
OK!

LPTK committed:
Make small fixes

LPTK committed:
Merge branch 'hkmc2' into hkmc2-companion-overloading

LPTK committed:
Merge remote-tracking branch 'CAG2Mark/lifter-companion-overloading' into hkmc2-companion-overloading

# Conflicts:
#	hkmc2/shared/src/test/mlscript/HkScratch.mls
#	hkmc2/shared/src/test/mlscript/lifter/ClassWithCompanion.mls

LPTK committed:
Merge remote-tracking branch 'origin/hkmc2' into hkmc2-companion-overloading

# Conflicts:
#	hkmc2/shared/src/test/mlscript/ucs/examples/EitherOrBoth.mls

LPTK committed:
Start cleanup

FlandiaYingman committed:
Improve WIP: Resolver and Overloaded Term|Type

LPTK committed:
Clean up JSBuilder

FlandiaYingman committed:
Support resolver with overloaded types

LPTK commented on hkmc2/shared/src/test/mlscript-compile/Str.mls:
```diff
@@ -1,5 +1,10 @@
+// The `object Str` in this file overrides the default `class Str`,
+// rather than overloading it. As a workaround, declare an overloaded
+// `class Str` again.
+declare class Str
```
Wow that's vile

LPTK replied:
Top-quality:
```fs
import "../../mlscript-compile/Str.mls"

:sjs
"a" is Str
//│ JS (unsanitized):
//│ let scrut; scrut = "a"; if (scrut instanceof globalThis.Str) { true } else { false }
//│ FAILURE: Unexpected runtime error
//│ FAILURE LOCATION: mkQuery (JSBackendDiffMaker.scala:154)
//│ ═══[RUNTIME ERROR] Error: Access to required field 'Str' yielded 'undefined'
```

FlandiaYingman replied:
Argh, how can we fix it?

FlandiaYingman replied:
How about we special-case the MLscript types `Str` `Num` etc. that will never be overridden by another definition, but will be overloaded with any later definition seen?

LPTK replied:
Actually how about we just don't mess with `Str`? 😬 

I think adding this module was a mistake. Could easily rename it to `StrOps` or just integrate the contents into `Predef`.

At some point we'll want to change the way `declare` works so we can write something like `declare class std.Str` – so we know this is a proper forward declaration and everything can actually fall back into place.

FlandiaYingman replied:
Okay, let's rename it to StrOps!

LPTK commented on hkmc2/shared/src/test/mlscript/basics/GenericClasses.mls:
```diff
@@ -41,7 +41,7 @@ class Bar(x: Foo[Int])
 //│ ║  l.39: 	class Foo[A](x: Bar[A])
 //│ ╙──      	                ^^^^^^
-//│ ╔══[ERROR] Expected 0 type arguments, got 1
+//│ ╔══[ERROR] Expected no type arguments, got 1
```
The most important change in this PR.

FlandiaYingman replied:
Amazing!

LPTK commented on hkmc2/shared/src/main/scala/hkmc2/semantics/Resolver.scala:
```diff
@@ -831,18 +832,93 @@ class Resolver(tl: TraceLogger)
+  def traverseType(t: Term, expect: Expect, inTyAppPrefix: Bool = false)(using ictx: ICtx): Unit =
+  trace(s"Traversing type ${t}, expecting ${expect}"):
+    def checkTypeArity(sym: FieldSymbol): Unit =
+      if inTyAppPrefix && !t.isInstanceOf[Term.TyApp] then
```
Why would you abort the checks in this case? Can you documemt the reason for the `!t.isInstanceOf[Term.TyApp]` check? @FlandiaYingman

FlandiaYingman replied:
Looks like a place that I forgot to clean up after changing the approach.

LPTK replied:
Ok, pls fix.

LPTK commented on hkmc2/shared/src/main/scala/hkmc2/semantics/Resolver.scala:
```diff
@@ -831,18 +832,93 @@ class Resolver(tl: TraceLogger)
+                msg"Expected ${tparamsMsg} type arguments, " +
+                msg"got ${targsMsg}" -> t.toLoc :: Nil
+    
+    def check(body: => Unit): Unit =
```
I still don't understand why you use this convoluted way of writing code. Why don't you just put the checking code after the traversing code, instead of adding a method with a by-name parameter? @FlandiaYingman

FlandiaYingman replied:
Oh, right LOL. I should have removed them. It was a legacy from the resolver prototype.

LPTK replied:
Ok, pls do remove them.

LPTK replied:
@FlandiaYingman So, will you remove them?

FlandiaYingman replied:
Yes!

LPTK committed:
Pass after Harry's changes

LPTK committed:
More cleanup

LPTK committed:
Start better modeling companions/overloading; but the new defn assigned to BlockMemberSymbol makes tests fail

LPTK committed:
Make small fixes

FlandiaYingman committed:
Clean up Harry's top-quality code

FlandiaYingman committed:
Local-dependent test diff

FlandiaYingman committed:
Change ModuleSymbol to ModuleOrObjectSymbol

FlandiaYingman committed:
Messy WIP Based on Term.Disamb

FlandiaYingman committed:
Disambiguiate by DisambBlockMemberSymbol

LPTK committed:
Fix toppest-quality code

LPTK committed:
Merge remote-tracking branch 'origin/hkmc2' into hkmc2-companion-overloading

# Conflicts:
#	hkmc2/shared/src/main/scala/hkmc2/codegen/Lowering.scala
#	hkmc2/shared/src/main/scala/hkmc2/syntax/Tree.scala
#	hkmc2/shared/src/test/mlscript/backlog/ToTriage.mls

FlandiaYingman committed:
Correctly expanding the top-quality `.class`

LPTK committed:
Apply Mark's suggestion

LPTK committed:
Finish cleaning up the code

LPTK commented on hkmc2/shared/src/main/scala/hkmc2/semantics/Term.scala:
```diff
@@ -44,79 +50,83 @@ sealed trait ResolvableImpl:
+    case S(N) => this
+    // case N => lastWords(s"missing expansion for term ${this}")
+    case N =>
+      // FIXME: instantiating a term without an expansion should be an
```
@FlandiaYingman Why? I thought `instantiate` would be what everyone would call by default before inspecting a term. (Remember this whole mutable business is just an optimization over just rewriting an immutable tree.)

FlandiaYingman replied:
I remember that comment was added before we agree on the change of expanding mechanism 😗 I'll do a cleanup for the resolver to remove all these misleading comments.

LPTK commented on hkmc2/shared/src/main/scala/hkmc2/semantics/Term.scala:
```diff
@@ -44,79 +50,83 @@ sealed trait ResolvableImpl:
   /** This method is only supposed to be called by Resolver. */
-  private[semantics] def expand(expansionFn: Opt[Term => Term]): this.type =
-    val newExpansion = expansionFn.map(_(t.duplicate.resolve))
+  private[semantics] def expand(expansionFn: Opt[this.type => Term]): this.type =
```
@FlandiaYingman Why does `expand` take a function?

FlandiaYingman replied:
That's a safety measure that I initially took to prevent from forgetting `duplicate`ing the term. Now it seems that it doesn't add much safety... I think we should remove it and accept `Term` instead.

LPTK replied:
Ok, pls make the change.

LPTK commented on hkmc2/shared/src/main/scala/hkmc2/semantics/Term.scala:
```diff
@@ -44,79 +50,83 @@ sealed trait ResolvableImpl:
-    case S(sym: MemberSymbol[?]) => sym.defn
+  def defn: Opt[Definition] = resolvedSymbol match
+    case S(sym: MemberSymbol[?]) => sym.defn.map(_.asPrincipalDefn)
+    // TODO: @Harry check if this case is really needed.
```
@FlandiaYingman 👀

FlandiaYingman replied:
OK I'll check it.

LPTK committed:
Remove bottom-quality hacks

LPTK commented on hkmc2/shared/src/main/scala/hkmc2/codegen/Lowering.scala:
```diff
@@ -282,23 +282,58 @@ class Lowering()(using Config, TL, Raise, State, Ctx):
+      // Q: is this ever supposed to be reached?
+      args(fs)(args => k(Value.Arr(mut = false, args)))
+    case ref @ st.Ref(sym) =>
+      // FIXME: DisambBlockMemberSymbol workaround
```
@FlandiaYingman Could you pls be more specific about what should be fixed, here?

FlandiaYingman replied:
Ideally after resolution, all `BlockMemberSymbol` should be resolved to `DisambBlockMemberSymbol`, so we should simply remove all cases about `BlockMemberSymbol` in the stages after resolution (i.e., `Lowering` and `Lifter`).

LPTK replied:
Ok, I meant please add that explanation to the comment! ;^)

LPTK committed:
Improve some tests and messages

LPTK committed:
Clean up the tests

LPTK commented:
Ok, the PR is ready. @FlandiaYingman @CAG2Mark please look into the agreed parts (and check your TODOs here: https://github.com/hkust-taco/mlscript/pull/331#issue-3348322909), and let me know if you see anything suspicious (also feel free to do a full review if you want).

FlandiaYingman committed:
Rename utility module Str to StrOps

FlandiaYingman commented:
Aye, aye, Captain

LPTK committed:
Propagate locs to JSBuilder's getVar to fix "missing definition" errors

Add `Term.clone` to prevent cross-origin `Loc` mix-up
Set `Loc` for selection name in `Compiler.reference`
Add missing `Loc` assignments

Co-authored-by: Luyu Cheng <2239547+chengluyu@users.noreply.github.com>

FlandiaYingman committed:
Refactor "convoluted" `check`

FlandiaYingman committed:
Cleanup Resolver and Resolvable 1

CAG2Mark commented:
Is there any reason `DisambBlockMemberSymbol` doesn't extend `BlockMemberSymbol`? I feel like it would make a lot of things cleaner.

LPTK commented on hkmc2/shared/src/main/scala/hkmc2/semantics/Term.scala:
```diff
@@ -58,19 +58,35 @@ sealed trait ResolvableImpl:
+
+  /** 
+   * Expanding a term to another, which can be later retrieved by the
+   * `instantiate` method. 
```
BTW this should probably be called `expanded`, not `instantiate`. There is no "instantiation" going on, here.

LPTK commented on hkmc2/shared/src/main/scala/hkmc2/codegen/Lifter.scala:
```diff
@@ -838,7 +838,7 @@ class Lifter(handlerPaths: Opt[HandlerPaths])(using State, Raise):
           .replCapturePaths(newCapturePaths)
           .replLocalPaths(newLocalsPaths)
-          .addIsymPaths(newIsymPaths)
+          .addIsymPaths(newIsymPaths.asInstanceOf) // FIXME
```
👀

LPTK commented on hkmc2/shared/src/main/scala/hkmc2/codegen/Lowering.scala:
```diff
@@ -57,7 +57,11 @@ import Subst.subst
   
   extension (t: Term)
-    def instantiated = t match
+    def instantiated =
```
```suggestion
    def expanded =
```

LPTK commented on hkmc2/shared/src/main/scala/hkmc2/codegen/Lowering.scala:
```diff
@@ -373,19 +377,20 @@ class Lowering()(using Config, TL, Raise, State, Ctx):
-          case bs: BlockMemberSymbol => bs -> bs.defn
-          case dbs: DisambBlockMemberSymbol[?] => dbs.bsym -> dbs.sym.defn
+        val (dbs, defn) = sym match
+          // FIXME: BMS legacy
```
👀 

(Just adding comments so we can keep track of what needs to be addressed in the PR before it can be merged.)

LPTK commented on hkmc2/shared/src/main/scala/hkmc2/semantics/Elaborator.scala:
```diff
@@ -219,9 +219,9 @@ object Elaborator:
-    val nonLocalRet =
-      val id = new Ident("ret")
-      BlockMemberSymbol(id.name, Nil, true)
+    // val nonLocalRet =
```
👀

LPTK commented:
Ok @FlandiaYingman the code of this PR was already cleaned up. I don't think it's a good idea to push to this branch commits with tons of comments and WIP changes that fail the tests. Better make keep these chnages separate for now and make a (possibly draft) PR to this branch.

LPTK commented on hkmc2/shared/src/main/scala/hkmc2/semantics/Symbol.scala:
```diff
@@ -201,15 +206,33 @@ class BlockMemberSymbol(val nme: Str, val trees: Ls[TypeOrTermDef], val nameIsMe
+    // case cls: ModuleOrObjectDef => cls.companion.isEmpty
+    // case _ => true
+
+  def asPrincipal: DefinitionSymbol[?] =
```
I thought we said we don't need this after all?

LPTK commented on hkmc2/shared/src/main/scala/hkmc2/syntax/Tree.scala:
```diff
@@ -222,7 +225,7 @@ enum Tree extends AutoLocated:
   def showDbg: Str = toString // TODO
   
-  lazy val desugared: Tree = this match
+  def desugared(using State): Tree = this match
```
WOW you can't just do that! The desugared tree is often accessed many time and shoudl NOT be recomputed each time!

FlandiaYingman committed:
Rename `instantiate` to `expanded`

FlandiaYingman committed:
”Remove" annoying `Unexpanded`

FlandiaYingman committed:
Remove annoying principal overload thingy

FlandiaYingman committed:
Remove annoying DisambBMS

FlandiaYingman committed:
Revert the revert

LPTK commented:
@FlandiaYingman I'm afrait your revert revert needs to be revereted, as it does not pass the CI.

FlandiaYingman committed:
Fix line number
